### PR TITLE
Fix warnings

### PR DESCRIPTION
**Pull request by hirocaster**
Fix <del>all</del> warnings.

### CODE REVIEW

File: lib/identity_cache.rb [hunk 1]
```diff
@@ -58,7 +58,7 @@ def included(base) #:nodoc:
     # +cache_adaptor+ - A ActiveSupport::Cache::Store
     #
     def cache_backend=(cache_adaptor)
-      if @cache
+      if instance_variable_defined?("@cache")
```
Comment (dylanahsmith) [inline lib/identity_cache.rb]: `if defined?(@cache)` is faster and more succinct

File: lib/identity_cache/configuration_dsl.rb [hunk 2]
```diff
@@ -205,11 +205,20 @@ def build_recursive_association_cache(association, options) #:nodoc:
...
         options[:records_variable_name]  = "cached_#{association}"
 
-        self.class_eval(<<-CODE, __FILE__, __LINE__ + 1)
-          def #{options[:cached_accessor_name]}
-            fetch_recursively_cached_association('#{options[:records_variable_name]}', :#{association})
-          end
-        CODE
+        if method_defined?(options[:cached_accessor_name].to_sym)
+          self.class_eval(<<-CODE, __FILE__, __LINE__ + 1)
+            def #{options[:cached_accessor_name]}_with_idc
```
Comment (dylanahsmith) [inline lib/identity_cache/configuration_dsl.rb]: why is this needed?

File: lib/identity_cache/query_api.rb [hunk 3]
```diff
@@ -399,10 +405,12 @@ def fetch_recursively_cached_association(ivar_name, association_name) # :nodoc:
...
         ivar_full_name = :"@#{ivar_name}"
 
-        unless ivar_value = instance_variable_get(ivar_full_name)
-          ivar_value = IdentityCache.map_cached_nil_for(send(association_name))
-          instance_variable_set(ivar_full_name, ivar_value)
-        end
+        ivar_value = if instance_variable_defined?(ivar_full_name)
+                       instance_variable_get(ivar_full_name)
+                     else
+                       value = IdentityCache.map_cached_nil_for(send(association_name))
```
Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: `IdentityCache.map_cached_nil_for` was to distinguish between `nil` and unloaded.  I think we should instead use the instance variable defined check to distinguish between these two states.  Since that isn't an isolated change, it should probably be handled in a separate PR.

File: lib/identity_cache/query_api.rb [hunk 4]
```diff
@@ -344,8 +350,8 @@ def prefetch_one_association(association, records)
               parent_record_to_child_records[parent_record] << child_record
             end
 
-            parent_record_to_child_records.each do |parent_record, child_records|
-              parent_record.send(details[:prepopulate_method_name], child_records)
+            parent_record_to_child_records.each do |parent, child|
```
Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: I think `children` is more appropriate than `child` since it is an array.

File: lib/identity_cache/query_api.rb [hunk 5]
```diff
@@ -319,7 +324,8 @@ def prefetch_embedded_association(records, association, details)
         first_record = records.first
         return if first_record.association(association).loaded?
         iv_name_key = details[:embed] == true ? :records_variable_name : :ids_variable_name
-        return if first_record.instance_variable_get(:"@#{details[iv_name_key]}")
+        iv_name = :"@#{details[iv_name_key]}"
+        return if first_record.instance_variable_defined?(iv_name) && first_record.instance_variable_get(iv_name)
```
Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: I think we only need to check if the instance variable exists

File: test/recursive_denormalized_has_many_test.rb [hunk 6]
```diff
@@ -61,7 +61,7 @@ def test_on_cache_miss_child_record_fetch_should_include_nested_associations_to_
       # one for the mid level has_one association
       # one for the deep level level has_many on the mid level has_many association
       # one for the deep level level has_many on the mid level has_one association
-      record_from_cache_miss = Item.fetch(@record.id)
+      Item.fetch(@record.id) # record_from_cache_miss
```
Comment (dylanahsmith) [inline test/recursive_denormalized_has_many_test.rb]: I think `# cache miss` would be a more appropriate comment.  Although I don't think a cache miss comment is necessary since the test name makes it clear that it is testing a cache miss.

File: test/recursive_denormalized_has_many_test.rb [hunk 7]
```diff
@@ -103,7 +103,7 @@ def setup
 
   def test_cache_repopulation_should_not_fetch_non_embedded_associations
     Item.any_instance.expects(:fetch_associated_records).never
-    record_from_cache_miss = Item.fetch(@record.id)
+    Item.fetch(@record.id) # record_from_cache_miss
```
Comment (dylanahsmith) [inline test/recursive_denormalized_has_many_test.rb]: I think `# cache miss` would be appropriate here

File: test/test_helper.rb [hunk 8]
```diff
@@ -11,7 +11,6 @@
 require File.dirname(__FILE__) + '/../lib/identity_cache'
 
 DatabaseConnection.setup
-ActiveSupport::Cache::Store.instrument = true
```
Comment (dylanahsmith) [inline test/test_helper.rb]: I think this is causing test failures.  E.g. `0 instead of 1 memcache operations were executed` I think is because it needs this instrumentation to count memcache operations

File: lib/identity_cache/configuration_dsl.rb (hunk 2 above)
Comment (hirocaster) [inline lib/identity_cache/configuration_dsl.rb]: this warning is `warning: method redefined; discarding old fetch_polymorphic_records`

Exists `#fetch_polymorphic_records` by AR.
`self.class_eval` is `#fetch_polymorphic_records` redefined.

This approach 

Define method(suffix `with_idc`), if exists method.
It using by alias_method_chain.

File: lib/identity_cache/query_api.rb (hunk 3 above)
Comment (hirocaster) [inline lib/identity_cache/query_api.rb]: OK, these changes remove in PR.
I will track #266.

File: test/recursive_denormalized_has_many_test.rb (hunk 6 above)
Comment (hirocaster) [inline test/recursive_denormalized_has_many_test.rb]: I think so too. Therefore remove cache miss comment.

**Comment by hirocaster** [comment]
Thank you code review, good speedy feedback. 
Fixed, please confirm.

File: lib/identity_cache/configuration_dsl.rb (hunk 2 above)
Comment (dylanahsmith) [inline lib/identity_cache/configuration_dsl.rb]: Active record doesn't define fetch_polymorphic_records.  It looks like this is coming from SchemaChangeTest#test_embed_existing_cache_has_many which is making sure it is safe to change an association between being fully embedded or just embedding the id.  This warning is an indication of a problem in the tests, not the code under test.

Perhaps that can be done more cleanly using

```ruby
teardown_models
setup_models
```

which is done between tests.

File: lib/identity_cache/query_api.rb [hunk 9]
```diff
@@ -182,7 +182,12 @@ def add_cached_associations_to_coder(record, coder)
           end
           if (cached_has_manys = klass.cached_has_manys).present?
             coder[:association_ids] = cached_has_manys.each_with_object({}) do |(name, options), hash|
-              hash[name] = record.instance_variable_get(:"@#{options[:ids_variable_name]}") unless options[:embed] == true
+              unless options[:embed] == true
+                iv_name = :"@#{options[:ids_variable_name]}"
+                hash[name] = if record.instance_variable_defined?(iv_name)
```
Comment (dylanahsmith) [inline lib/identity_cache/query_api.rb]: It looks like this check is unnecessary, and was just added to workaround a warning in another specific test. SchemaChangeTest#test_add_non_embedded_cache_has_many does `Item.expects(:resolve_cache_miss).returns(@record)` which mocks out the code that sets this variable.

File: lib/identity_cache/configuration_dsl.rb (hunk 2 above)
Comment (hirocaster) [inline lib/identity_cache/configuration_dsl.rb]: Thank you, comment. :smile: 

I understand define fetch_polymorphic_records. 
I red

- ActiveRecordObjects
- #build_recursive_association_cache
- #build_id_embedded_has_many_cache

:eyes: 
--

Thats is certainly redefin #fetch_polymorphic_records in SchemaChangeTest#test_embed_existing_cache_has_many.

File: test/schema_change_test.rb [hunk 10]
```diff
@@ -89,10 +89,12 @@ def test_embed_existing_cache_has_many
 
     record = Item.fetch(@record.id)
 
+    teardown_models
+    setup_models
```
Comment (hirocaster) [inline test/schema_change_test.rb]: I think not best. :disappointed:

**Comment by dylanahsmith** [comment]
Thanks for the fixes
